JT/12_overloading_overriding/overloading_vs_overriding.adoc

```diff
@@ -0,0 +1,641 @@
+
+
+* **Type Promotion:**
+Java uses type promotion for matching arguments, and this can lead to unexpected results. For example, if you have overloaded methods where one takes a long and another takes an int, and you call the method with a byte value, the version with the int parameter will be called, not the long.
+
+
+* **Testing Difficulty:**
+Overloaded methods can make writing unit tests more difficult, as you need to ensure that all versions of the method are correctly tested.
+
+link:#table-of-content[Back to top]
+
+## [[conclusion]]Method Overloading Rules:
+
+There are some rules which we need to follow to overload a method and some of them are mandatory while some are optional.
+
+Two methods will be treated as Overloaded if both follow below mandatory rule.
+
+- Both must have same method name
+
+- Both must have different argument lists
+
+And if both methods follow above mandatory rules then they may or may not
+
+- Have different return types
+
+- Have different access modifiers
+
+- Throw different checked or unchecked exceptions
+
+Usually, method overloading happens inside a single class but a method can also be treated as overloaded in the subclass of that class because subclass inherits one version of the method from the parent class and then can have another overloaded version in its class definition.
+
+link:#table-of-content[Back to top]
+
+
+## [[method-overload-practice-exercises]]Method overloading practice exercises:
+
+### Exercise 1:
+Create a class with a method that prints “ONE” when no arguments are passed, and “TWO” when one integer argument is passed.
+
+[%collapsible]
+.Solution
+====
+[source,java]
+----
+public class MainClass {
+    static void overloadedMethod() {
+        System.out.println("ONE");
+    }
+
+    static void overloadedMethod(int a) {
+        System.out.println("TWO");
+    }
+
+    public static void main(String[] args) {
+        overloadedMethod();
+        overloadedMethod(1);
+    }
+}
+----
+====
+
+
+### Exercise 2:
+In the below class, is ‘method’ overloaded or duplicated?
+
+[%collapsible]
+.Solution
+====
+[source,java]
+----
+public class MainClass {
+    void method(int... a) {
+        System.out.println(1);
+    }
+
+    void method(int[] a) {
+        System.out.println(2);
+    }
+}
+----
+====
+
+### Exercise 3:
+In the below Class X, is ‘method’ properly overloaded?
+
+[%collapsible]
+.Solution
+====
+[source,java]
+----
+class X {
+    int method(int i, int d) {
+        return i + d;
+    }
+
+    static int method(int i, double d) {
+        return (int) (i + d);
+    }
+
+    double method(double i, int d) {
+        return i + d;
+    }
+
+    static double method(double i, double d) {
+        return i + d;
+    }
+}
+----
+====
+
+### Exercise 4:
+Create a class with overloaded methods that accept different number of arguments and data types.
+
```

Comment:
What was meant here relating to data types - the different **number** of data types or that the **data types** themselves **should be different**? If it was the 2nd case then it would be nice to change the example.

Reply:
There are no logical errors in the given example. The `MethodOverloading` class contains two overloaded `display` methods that accept a different number of arguments of type `int`. 
In the `main` method, these methods are called with different numbers of arguments, demonstrating method overloading. Each method prints the arguments to the console, showing that overloading works correctly. The author uses the word "or" in the task. So, the condition is met.
For a more in-depth discussion, I suggest reaching out to the author of the text - @Tudor Mateevici.

Reply:
Fair enough. I have changed condition of task. We use the word "or" (instead 'and') in the task. So, the condition is met.
There are no logical errors in the given example. The MethodOverloading class contains two overloaded display methods that accept a different number of arguments of type int. In the main method, these methods are called with different numbers of arguments, demonstrating method overloading. Each method prints the arguments to the console, showing that overloading works correctly.